File: 11-minimum-parenthesis/20220830.ts
function minRemoveToMakeValid(s: string): string {
    const chars: string[] = s.split('');
    const unmatchedLeft: number[] = [];
    let position: number;

    for (let i = 0; i < chars.length; ++i) {
        if (chars[i] === '(') {
            // Remember the position
            unmatchedLeft.push(i);
        } else if (chars[i] === ')') {
            if (unmatchedLeft.length === 0) {
                // If no left parentheses, remove this right parenthesis.
                chars.splice(i, 1);
                --i;
            } else {
                // Remove a left parenthesis from the stack since we have a match.
                unmatchedLeft.pop();
            }
        }
    }

    // Remove any unmatched left parenthesis
    while((position = unmatchedLeft.pop()) !== undefined) {
        chars.splice(position, 1);
    }

    return chars.join('');
}

File: 11-minimum-parenthesis/20220830.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        s = list(s)
        unmatchedLeft = []

        for i in range(len(s) - 1):
            print(len(s) - 1)
            if s[i] == '(':
                unmatchedLeft.append(i)
            elif s[i] == ')':
                if len(unmatchedLeft) == 0:
                    del s[i]
                    i = i - 1
                else:
                    unmatchedLeft.pop()

        while len(unmatchedLeft):
            position = unmatchedLeft.pop()
            del s[position]

        return ''.join(s)


print(Solution().minRemoveToMakeValid('a(b)c)'))
